### ISSUE TITLE
Feature request: output C with intrinsics

### ISSUE THREAD

**Issue by andres-erbsen** [0xADE1A1DE/CryptOpt#145]
It'd be neat if we could have CryptOpt do optimization on assembly as it does, but then print the scheduled and register-allocated code as C to see how much worse C compilers do when given the easiest possible task. The code generated in this manner might also be preferrable to raw fiat-crypto output in cases a C implementation needs to be deployed. I don't think we should make any effort to use standard intrinsics, rather let's do something like fiat-crypto where we just define C functions that compute the same values as supported assembly instructions.

**Comment by dderjoel** [comment]
I'm not sure how register-allocated code as C would look like. Would I want to create  C symbols like `rax_1` ? But then the compiler can register allocate anything to then, right?
Another tricky point would be how to deal with flags, as I'm not aware of C-semantics for overflow/carry flags. 

The way it currently works it that I preprocess any input (in this case from the fiat-bridge). Then this becomes my internal representation which in combination with a RegisterAllocator can be written to an assembly string. 

In other words, there is currently no interface from my IR to 'write to x86-64'. 
This ties in with the #143 and possible other targets, to define some interface, from which then, given some capabilities, a string representation (Go, C-intrinsics, x86-64, arm, ...) can be emitted.

**Comment by andres-erbsen** [comment]
Yes, I was imagining C variables for both registers and flags, and custom C prototypes I would implement to emulate various flag-manipulating instructions. I don't think we need any capability-selection logic for this, just a different stringification procedure. I might be missing something about the post-regalloc passes in CryptOpt though, I have only barely looked at the code.

**Comment by andres-erbsen** [comment]
Yeah, I think I am coming around to thinking that maybe having this functionality be a part of cryptopt is not that good of an idea after all, for the reasons you discuss. I even looked up the `register` annotation in C and turns out it is incompatible with intrinsics whose outputs are passed by pointer.

Perhaps I'll instead try a sed script that translates every opcode in an assembly file to a similarly-named macro in C... it's not going to be pretty with sub-register access though, but maybe I'll get somewhere without supporting that.
